src/modules/utils/java_version_parser.rs

```diff
@@ -0,0 +1,61 @@
+use nom::{
+    branch::alt,
+    bytes::complete::{tag, take_until, take_while1},
+    combinator::rest,
+    sequence::{preceded, tuple},
+    IResult,
+};
+
+fn is_version(c: char) -> bool {
+    c >= '0' && c <= '9' || c == '.'
+}
+
+fn version(input: &str) -> IResult<&str, &str> {
+    take_while1(&is_version)(input)
+}
+
+fn zulu(input: &str) -> IResult<&str, &str> {
+    let zulu_prefix_value = preceded(take_until("("), tag("("));
+    preceded(zulu_prefix_value, version)(input)
+}
+
+fn jre_prefix(input: &str) -> IResult<&str, &str> {
+    preceded(take_until("JRE ("), tag("JRE ("))(input)
+}
+
+fn j9_prefix(input: &str) -> IResult<&str, &str> {
+    preceded(take_until("VM ("), tag("VM ("))(input)
+}
+
+fn suffix(input: &str) -> IResult<&str, &str> {
+    rest(input)
+}
+
+/// Parse the java version from `java -Xinternalversion` format.
+///
+/// The expected format is similar to:
+///     "JRE (1.8.0_222-b10)"
+///     "JRE (Zulu 8.40.0.25-CA-linux64) (1.8.0_222-b10)"
+///     "VM (1.8.0_222-b10)".
+///
+/// Some Java vendors might not follow this format.
+pub fn parse_jre_version(input: &str) -> IResult<&str, &str> {
+    let prefix = alt((jre_prefix, j9_prefix));
+    let version_or_zulu = alt((version, zulu));
+    let (input, (_, version, _)) = tuple((prefix, version_or_zulu, suffix))(input)?;
+
+    Ok((input, version))
```

Comment:
What's the reason for returning a tuple including the `input`?
It looks to me like we don't use it anywhere, and makes the output hard to reason with without reading through the implementation of `parse_jre_version`.

Reply:
You are right, the parser API should hide its internals. Now it doesn't expose the leftovers from `IResult`.